src/test/java/com/car/app/CarDaoTest.java

```diff
@@ -0,0 +1,123 @@
+package com.car.app;
+
+import static org.junit.Assert.assertEquals;
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.ArgumentMatchers.anyString;
+import static org.mockito.Mockito.doNothing;
+import static org.mockito.Mockito.when;
+
+import java.util.ArrayList;
+import java.util.Date;
+import java.util.List;
+
+import javax.persistence.EntityManager;
+import javax.persistence.EntityTransaction;
+import javax.persistence.Query;
+
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.mockito.InjectMocks;
+import org.mockito.Mock;
+import org.mockito.junit.MockitoJUnitRunner;
+
+import com.car.app.controllers.CarDao;
+import com.car.app.model.Car;
+
+@RunWith(MockitoJUnitRunner.class)
+public class CarDaoTest {
+
+	@InjectMocks
+	CarDao carDao;
+	
+	@Mock
+	EntityManager entityManager;
+	
+	@Mock
+	EntityTransaction transaction;
+	
+	@Mock
+	Query query;
+	
+	List<Car> cars;
+	
+	/**
+	 * Configuration before any test method
+	 */
+	@Before
+	public void init() {
+		
+		cars = new ArrayList<Car>();
+		for(int i = 0; i < 5; i++) {
+			Car car = new Car();
+			car.setBrand("Brand "+i);
+			car.setCountry("Country - "+i);
+			car.setRegistration(new Date());
+			car.setId("id-"+i);
+			cars.add(car);
+		}
+		
+		// mock the entity manager
+		when(entityManager.getTransaction()).thenReturn(transaction);
+		doNothing().when(transaction).begin();
+		doNothing().when(transaction).commit();
+		doNothing().when(entityManager).clear();
+	}
+	
+	/**
+	 *  Testing the method findAllCars
+	 */
+	@Test
+	public void testFindAllCars() {
+		
+		// mock the query:
+		when(entityManager.createQuery(anyString())).thenReturn(query);
+		when(query.getResultList()).thenReturn(cars);
+		
+		List<Car> cars2 = carDao.findAllCars();
+		assertEquals(cars2.size(), cars.size());
```

Comment:
How is this test ensuring that the query executed is the correct one ?

Reply:
yeah! I see. 
so we have to make real queries to db ?

Reply:
No, not really, you can use Mockito ArgumentCaptor : 
https://www.baeldung.com/mockito-argumentcaptor